### PR TITLE
feat: Adding SplitDB and refactoring ColdDB

### PR DESCRIPTION
**Pull request by wacban**
Note for reviewers: commits are mostly self-contained so this PR can be reviewed commit by commit if that's easier. 

- Removed the generic D from ColdDB and NodeStorage. Instead just use the existing Database trait. It's equally generic and we can still use some non-RocksDB implementation down the line but it's much cleaner and in line with rest of the codebase. 
- Removed the hot database from ColdDB. The ColdDB should now only be used for accessing the cold storage and be responsible for the different format of data in the cold storage.
- Added SplitDB that implements the read hot first, read cold second approach to reading from the databases. 

This is in preparation for having the ViewClient use the SplitDB as means for reading from cold storage. 

The ultimate goal is to have something like the following
```
NodeStorage {
  hot: Arc<Database>;    // typically RocksDB
  cold: Arc<Database>;   // typically ColdDB with RocksDB inside of it
  split: Arc<Database>;  // typically SplitDB with both hot and cold inside of it. This may be a field or there may just be a getter for it that constructs it on the fly. The important bit is that it's possible to get the split storage from the NodeStorage. 
}

// Responsible for adjusting the format to cold - get rid of rc and a few other.
// Has limited functionality and can only handle DbVersion, BlockMisc and the cold columns.
// Can be used e.g. when populating the cold storage in the cold store loop.
struct ColdDB {
  db: Arc<Database> 
}
impl Database for ColdDB {
...
}

// Responsible for implementing the "try hot first, try cold later" approach.
// Has full functionality and can handle all columns.
// Can be used as the storage backing Store e.g. in view client.
struct SplitDB {
  hot: Arc<Database>;
  cold: Arc<Database>;
}

impl Database for SplitDB {
...
}
```

### CODE REVIEW

File: core/store/src/db/splitdb.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    /// For cold columns returns the cold iterator.
+    /// For other columns returns the hot iterator.
+    ///
+    /// Keep in mind that the cold db is typically behind by a few blocks. It
+    /// may be behind by up to gc_num_epochs_to_keep epochs under acceptable
+    /// conditions or more if the cold store loop is falling behind under
+    /// erroneous conditions.
+    fn iter<'a>(&'a self, column: DBCol) -> DBIterator<'a> {
+        match column.is_cold() {
+            false => self.hot.iter(column),
```
Comment (nikurt) [inline core/store/src/db/splitdb.rs:86]: The comment states that cold db is typically behind.
Is it acceptable to iterate only over cold db? Should the iteration be done over both cold and hot?

File: nearcore/src/cold_storage.rs [hunk 2]
```diff
@@ -185,14 +185,14 @@ pub fn spawn_cold_store_loop(
     let cold_store = match storage.get_cold_store() {
         Some(cold_store) => cold_store,
         None => {
-            tracing::debug!(target:"cold_store", "Not spawning cold store loop because cold store is not configured");
+            tracing::debug!(target : "cold_store", "Not spawning the cold store loop because cold store is not configured");
```
Comment (nikurt) [inline nearcore/src/cold_storage.rs:188]:
```suggestion
            tracing::debug!(target: "cold_store", "Not spawning the cold store loop because cold store is not configured");
```

File: nearcore/src/cold_storage.rs [hunk 3]
```diff
@@ -185,14 +185,14 @@ pub fn spawn_cold_store_loop(
...
             return Ok(None);
         }
     };
     let cold_db = match storage.cold_db() {
         Some(cold_db) => cold_db.clone(),
         None => {
-            tracing::debug!(target:"cold_store", "Not spawning cold store loop because cold store is not configured");
+            tracing::debug!(target:"cold_store", "Not spawning the cold store loop because cold store is not configured");
```
Comment (nikurt) [inline nearcore/src/cold_storage.rs:195]:
```suggestion
            tracing::debug!(target: "cold_store", "Not spawning the cold store loop because cold store is not configured");
```

**Review by nikurt — APPROVED** [review]
Looks great

File: core/store/src/db/splitdb.rs (hunk 1 above)
Comment (wacban) [reply core/store/src/db/splitdb.rs:86]: Great question but I don't have the answer. In this PR I wanted to get a minimal sensible implementation so I did it this way. For hot only columns it is correct to only iterate over hot. For cold columns we may want iterate over both but then there is the issue of duplicated entries. To make things complicated when copying data to cold, for some columns, we change the format of the key. IIRC for some column we remove shard id from the key and for some other we change endianness (don't ask me why). This would change the order of iteration and make duplicate removal hard. 

TLDR: let's keep it as is for now and iterate (pun intended) on this implementation as needed.

**Comment by wacban** [comment]
https://nayduck.near.org/#/run/2881
